mithril-core/src/stm.rs

```diff
@@ -200,30 +204,36 @@ pub struct StmSig<D: Clone + Digest> {
     pub stake: Stake,
     /// The index(es) for which the signature is valid
     pub indexes: Vec<Index>,
-    /// The path through the MerkleTree for this party.
-    pub path: Path<D>,
+    /// Merkle tree index of the signer.
+    pub signer_index: Option<Index>,
+    hasher: PhantomData<D>,
+
 }
 
-/// Stm aggregate key, which contains the merkle tree root and the total stake of the system.
+/// Stm aggregate key (batch compatible), which contains the merkle tree commitment and the total stake of the system.
+/// Batch Compat Merkle tree commitment includes the number of leaves in the tree in order to obtain batch path.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(bound(
-    serialize = "Path<D>: Serialize",
-    deserialize = "Path<D>: Deserialize<'de>"
+    serialize = "BatchPath<D>: Serialize",
+    deserialize = "BatchPath<D>: Deserialize<'de>"
 ))]
-pub struct StmAggrVerificationKey<D: Clone + Digest> {
+pub struct StmAggrVerificationKey<D: Clone + Digest + FixedOutput> {
     mt_commitment: MerkleTreeCommitment<D>,
     total_stake: Stake,
 }
 
 /// `StmMultiSig` uses the "concatenation" proving system (as described in Section 4.3 of the original paper.)
 /// This means that the aggregated signature contains a vector with all individual signatures.
+/// BatchPath is added as `Option` for batch compatibility.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(bound(
-    serialize = "Path<D>: Serialize",
-    deserialize = "Path<D>: Deserialize<'de>"
+    serialize = "BatchPath<D>: Serialize",
+    deserialize = "BatchPath<D>: Deserialize<'de>"
 ))]
-pub struct StmAggrSig<D: Clone + Digest> {
+pub struct StmAggrSig<D: Clone + Digest + FixedOutput> {
     pub(crate) signatures: Vec<StmSig<D>>,
+    /// The list of unique merkle tree nodes that covers path for all signatures.
+    pub batch_proof: Option<BatchPath<D>>,
```

Comment:
Do we need this to be an option? Do we accept an `StmAggrSig` without a batch_proof?